conda_smithy/lint_recipe.py

```diff
@@ -186,7 +186,10 @@ def lintify_meta_yaml(
     lint_build_section_should_be_before_run(requirements_section, lints)
 
     # 9: Files downloaded should have a hash.
-    lint_sources_should_have_hash(sources_section, lints)
+    if recipe_version == 1:
+        lint_sources_should_have_hash(sources_section, lints)
+    elif recipe_version == 2:
+        conda_recipe_v2_linter.lint_sources(sources_section, lints)
```

Comment:
There's no need to have two functions. Please use one.

Reply:
Do you want to lint v1 recipes to make sure that the SHA hash is not templated as well @isuruf?
I think that would be a good change, but some churn involved.

Reply:
You can pass `recipe_version` to `lint_sources_should_have_hash` and make the distinction there. Btw, how do you write a recipe like https://github.com/conda-forge/cuda-nvcc-impl-feedstock/blob/main/recipe/meta.yaml#L24-L28?

Reply:
You can write it like this (valid yaml!):

```yaml
context:
  version: "12.6.20"

source:
  - if: target_platform == "linux-64:
    then:
      url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/linux-x86_64/cuda_nvcc-linux-x86_64-{{ version }}-archive.tar.xz
      sha256: 4336a230269db14336ca1f727a0502631390a52d37c17d3bde05ccda5d534406
      patches:
        - nvcc.profile.patch
  - if: target_platform == "linux-aarch64"
    then:
      url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/linux-sbsa/cuda_nvcc-linux-sbsa-{{ version }}-archive.tar.xz
      sha256: 06112bf1cc0cfb6b881a2e0420cb4611b6d701decbb195fbfb4f0e9966a42006
      patches:
        - nvcc.profile.patch
  - if: target_platform == "win-64"
    then:
      url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/x64/cuda_nvcc-x64-{{ version }}-archive.zip
      sha256: 03282a36fb98c2f227877a62efd04c36b1ab01fb1b676d9edbd0b10179e0a40a
      patches:
        - nvcc.profile.patch.win
```

Reply:
biggest downside is the `patch` duplication for the two sources although I think that's ... OK :)

Reply:
We should brainstorm ideas on how to prevent the duplication there.

Reply:
Yeah we should allow templates and write bot code that properly updates the recipe. This entire process was designed to eliminate an edge case in the bot code base for v2 recipes and it doesn't actually work for that in general.

Reply:
@beckermr I think this is a side-discussion :)

I don't really mind that little duplication (and it only happens for this kind of binary repackaging). 

For other recipes, you can do things like:

```
url: foo
patches:
- ${{ "linux.patch" if linux }}
- ${{ "win.patch" if win }}
```

But if you come up with something clever thats not intrusive, I am all ears. I just think this is really a special case recipe (compared to the majority of straight-forward single source recipes).

Reply:
The entire discussion is about special cases which this lint was supposed to eliminate. If it doesn't actually do that, then we're simply excluding part of a new recipe spec (that we created!) because we don't want to deal with the implications. That would indicate to me that either we need to deal with the special cases or we need to reconsider/modify the v2 recipe spec. It is strange to me for us to create a spec with some feature and then immediately exclude it.

Reply:
> All in all I think the new source section is a big improvement over the previous linked one, just because it's valid YAML (no duplicate keys!).

We could have done with v1
```
source:
  - url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/linux-x86_64/cuda_nvcc-linux-x86_64-{{ version }}-archive.tar.xz  # [linux64]
    sha256: 4336a230269db14336ca1f727a0502631390a52d37c17d3bde05ccda5d534406  # [linux64]
    patches:  # [linux64]
      - nvcc.profile.patch # [linux64]
  - url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/linux-sbsa/cuda_nvcc-linux-sbsa-{{ version }}-archive.tar.xz  # [linux and aarch64]
    sha256: 06112bf1cc0cfb6b881a2e0420cb4611b6d701decbb195fbfb4f0e9966a42006 # [linux and aarch64]
    patches:  # [linux and aarch64]
        - nvcc.profile.patch  # [linux and aarch64]
  - url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/x64/cuda_nvcc-x64-{{ version }}-archive.zip  # [win64]
    sha256: 03282a36fb98c2f227877a62efd04c36b1ab01fb1b676d9edbd0b10179e0a40a  # [win64]
    patches:    # [win64]
        - nvcc.profile.patch.win   # [win64]
```
which is valid YAML ((no duplicate keys!)
So, it's not really an improvement. I think if you want to ban templating in the hash, then we should have an alternative. As @beckermr said this leads to a deficiency in v2 spec.

Reply:
I don't get these points at all. sorry. 

The current implementation of the version update bot is pretty strict. 
It requires:

- the build number to be a variable that is either `build = `, `build_number =` or to be directly set in the build section
- the version to be in a variable that is named _exactly_ version.

But now you are arguing wether we should make it also strict about having the hash in the source section vs. as a variable (that has _no_ obvious benefit).

I am solely proposing to add a "conda-forge" specific lint to make sure that the hash is not templated. I am proposing that to simplify the work for the version update bot. I haven't heard a good argument for templating the hash.
It _is_ possible to template the hash - nothing prevents you from doing so if you like it in the v2 spec / rattler-build (only our current pydantic recipe-schema implementation will give you red squiggles because it uses a regex to detect a SHA256 hash, but we can fix that, too).

Of course you can do whatever you like in the V1 recipe. That's what we're trying to fix. You can probably write turing complete programs with it. I don't see the point you are trying to make here :)

We also have a lint to have a extra empty line at the end of the file ... to make the life of the bots easier. And that's a pretty annoying lint in my book! :)

Reply:
You are making functionality limitations to make the bot work and the only way to workaround that is to do a bunch of duplication (for eg: duplicating `patches` here). Imposing a functionality limitation is the issue here. 

> We also have a lint to have a extra empty line at the end of the file ... to make the life of the bots easier. And that's a pretty annoying lint in my book! :)

This is not a functionality limitation, so they are vastly different.

Reply:
Well in the v2 recipe format, what you want to do is annoying to write, one way or the other. So I do not see any functionality limitation.

If I wanted to try to translate your recipe:

```yaml
context:
  platform: ${{ "linux-x86_64" if linux and x86_64 else "linux-ppc64le" if ppc64le else "linux-sbsa" if aarch64 else "windows-x86_64" }}
  sha256: ${{ "4336a230269db14336ca1f727a0502631390a52d37c17d3bde05ccda5d534406" if linux and x86_64 else "06112bf1cc0cfb6b881a2e0420cb4611b6d701decbb195fbfb4f0e9966a42006" if linux and aarch64 else "03282a36fb98c2f227877a62efd04c36b1ab01fb1b676d9edbd0b10179e0a40a" if win }}
  extension: ${{ "tar.xz" if unix else "zip" }}

source:
  url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_nvcc/{{ platform }}/cuda_nvcc-{{ platform }}-{{ version }}-archive.{{ extension }}
  sha256: ${{ sha256 }}
  patches:
    - ${{ "nvcc.profile.patch" if linux }}
    - ${{ "nvcc.profile.patch.win" if win }}
```

I think a lint to prevent such abominations is justified :)

Reply:
Since you agree that is an abomination, we should figure out a non-abomination. That's the whole point of v2 IMO. Make recipes not be abominations.

Reply:
So for context, the lint suggestion arose here

https://github.com/regro/cf-scripts/pull/2920#discussion_r1709527413 

where I commented that you'd need to update the hash in the context section too. The suggestion was that by eliminating the use of variables to specify the hash, we could simplify the code in the bot. That entire block of code is 24 lines in the bot (https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/update_recipe/version.py#L173-L197). 

Then we had another discussion here

https://github.com/regro/cf-scripts/pull/2920/files/b7f61366a355ab41e0728e7cc325ed36596c3d67#discussion_r1709740255

where I pointed out that we have to evaluate the context section of the recipe under the assumptions of not just linux, but also other platforms in order to update things correctly. You suggested that this lint would help and we both agreed there would be other edge cases this would miss. 

So we can have a separate discussion on this change on its own, but the context for it came from simplifying bot code by eliminating flexibility in the v2 recipe spec.

This is the motivation for my comments. This PR is an incomplete solution to solve complexity in the v2 recipe spec. We either need to reconsider the spec to eliminate complexity or we need to deal with it. Excluding the a small part of the complexity only in conda-forge is an incomplete solution. Given that many conda-forge devs helped make the spec, it's rather odd that we wouldn't use all of it. 

As for the bot and the implementation, I don't see the point in avoiding the work to support the new spec in the same way. Otherwise, we do something that covers only part of the cases, people find bugs + submit bug reports, and then someone (you, me, someone else?) in the future is left either saying this won't ever happen or fixing it. 

I suggested a better path forward for this work in multiple places related to eliminating technical debt around recipe parsing+editing (which is one of the original motivations for the new spec anyways) and then using that to move things ahead.

Reply:
Hmm, I think my motivation was never to make it less verbose to write complex recipes :)

My motivation was to make recipes structured, easy to parse and fast to evaluate (ie. not do the recursive craziness of conda-build).

I do think we could find solutions to this problem, but for me it's low priority as we can continue using meta.yaml, so if a user doesn't like the new format / additional verbosity there is no need to switch. From my POV it's also very few recipes that are structured this way (mostly binary repacks). And I also don't mind the more extended form, personally.

But if you have good ideas for improvements to the spec, I would love to hear them!

Thanks for adding the context, Matt. I am happy to fix bugs/issues as they come up related to the new format, of course!

Reply:
We're talking past one another a bit here, but this statement is a red flag:

> I do think we could find solutions to this problem, but for me it's low priority as we can continue using meta.yaml, so if a user doesn't like the new format / additional verbosity there is no need to switch. From my POV it's also very few recipes that are structured this way (mostly binary repacks). And I also don't mind the more extended form, personally.

The goal here, I thought, is to move to a new spec and eliminate a previous one, all for tangible benefits (performance, interpretability, maintainability, etc.). If we are not in this process to actually deprecate and then remove the v1 spec, why on earth would we do all of this work just to support both in the end?

> Thanks for adding the context, Matt. I am happy to fix bugs/issues as they come up related to the new format, of course!

Great and thank you! I have raised several issues around edge cases in the PRs we have going for v2 and my opinion is that now is a good time to address them.

Reply:
If the outcome is that v2 replaces v1 I will be personally very happy. But I also don't want to force people on v2, especially not during the initial period. The whole transition will probably take at least a year, and the idea is that it's opt-in (if you have a recipe.yaml then new format is used, otherwise fall back to meta.yaml). I hope we're on the same page there :)